### ISSUE TITLE
%X:=timezone()% modifier not working as expected

### ISSUE THREAD

**Issue by ckeydel** [RealRaven2000/SmartTemplate4#5]
The time stamps in email headers always use "universal" time, e.g.

`Date: Wed, 7 Aug 2019 16:58:51 +0200`

The time here is the local (server) system time and the offset is the system's offset to UTC (==GMT). So in this example, the UTC time would be 14:58:51, which is what would be expected when using

` %X:=timezone(UTC)%`

In this case, %date_tz% should also output "0000" and %tz_name% should print "UTC".

It doesn't do that currently, though. I haven't figured out what calculation it uses, but I believe it has something to do with the client computer's time zone.

This is how it should work IMO:

- If there is no %X:=timezone()% modifier, default to %X:=timezone(\<local client time zone\>)%, %date_tz% its UTC offset and %tz_name%=\<local client time zone\>
- Calculate UTC from the time stamp in email header if possible. If that is not possible, assume TB got it right and **substract** local client time zone UTC offset to get "assumed UTC"
- If there is a %X:=timezone()% modifier, set %date_tz% and %tz_name% accordingly. Add this time zone's UTC offset to the previously calculated UTC time.

With this, the time will always be correct no matter in which time zone the user wants to output it.

**Comment by RealRaven2000** [comment]
To see what exactly happens you can enable [x] Debug mode on the advanced settings tab. by right-clicking the debug checkbox, one can also add the switches

```
extensions.smartTemplate4.debug.timeStrings = true
extensions.smartTemplate4.debug.timeZones = true
```
This will show additional protocol lines in Error Console.

 It attempts to display the timestamp in the "forced" time zone; e.g. if you add _%X:=timezone(BST)%_ then the sent times from the original email are expected to be expressed in that timezone.

```
if (SmartTemplate4.whatIsTimezone) {
  let forcedTz = util.getTimezoneOffset(SmartTemplate4.whatIsTimezone);
  msOffset = msOffset + forcedTz*60*60*1000 - nativeUtcOffset*60*1000;
  util.logDebug("Adding timezone offsets:\n" +
    "UTC Offset: " + nativeUtcOffset/(-60) + "\n" +
    "Forced Timezone:" + forcedTz);
}
```

The function to calculate the timezone offset  has a hardcoded list:

```
  getTimezoneOffset: function st4_getTimezoneOffset(zone) {
    let offset = 0;
    // Offsets according to https://en.wikipedia.org/wiki/List_of_time_zone_abbreviations
    switch (zone) {
      case "ACDT": return 10.5;
      case "ACST": return 9.5;
      case "ACT": return -5;
      case "ACWST": return 8.75;
      case "ADT": return -3;
      case "AEDT": return 11;
      case "AEST": return 10;
      case "AFT": return 4.5;
      case "ART": return -3;
      case "AKDT": return -8; // Alaska Daylight
      case "AKST": return -9;
      case "AMST": return -3;
      case "ART": return -3;
      case "AST": return -4;  //Atlantic Standard Time
      case "AWST": return 8;
      case "AZOST": return 0;
      case "AZOT": return -1;
      case "AZT": return 4;
      case "BDT": return 8;
      case "BIOT": return 6;
      case "BIT": return -12;
      case "BOT": return -4;
      case "BRST": return -2;
      case "BRT": return -3;
      case "BST": return 1; // British Summer Time
      case "BTT": return 6;
      case "CAT": return 2;
      case "CCT": return 6.5;
      case "CDT": return -5; // Central Daylight (North America)
      case "CEST": return 2;
      case "CET": return 1;
      case "CHADT": return 13.75;
      case "CHAST": return 12.75;
      case "CHOT": return 8;
      case "CHIST": return 9;
      case "CHST": return 10;
      case "CHUT": return 10;
      case "CIST": return -8;
      case "CIT": return 8;
      case "CKT": return -10;
      case "CLST": return -3;
      case "CLT": return -4;
      case "COST": return -5;
      case "CST": return -6; // Central Standard Time (North America)
      case "CT": return 8;
      case "CVT": return -1;
      case "CWST": return -8.75; // Central Western Standard Time (Australia) unofficial
      case "CXT": return 7;
      case "DAVY": return 7;
      case "DDUT": return 10;
      case "DFT": return 1;
      case "EASST": return -5;
      case "EAST": return -6;
      case "EAT": return 3;
      case "ECT": return -4; //Eastern Caribbean Time 
      case "EDT": return -4; // Eastern Daylight Time (North America)
      case "EEST": return 3;
      case "EET": return 2;
      case "EGST": return -1;
      case "EGT": return -1;
      case "EIT": return 9;
      case "EST": return -5; // Eastern Standard Time (North America)
      case "FET": return 3;
      case "FJT": return 12;
      case "FKST": return -3;
      case "FKT": return -4;
      case "FNT": return -2;
      case "GAKT": return -6;
      case "GAMT": return -9;
      case "GFT": return -3;
      case "GILT": return 12;
      case "GIT": return -9;
      case "GMT": return 0;
      case "GST": return 4; //    Gulf Standard Time
      case "GYT": return -4;
      case "HDT": return -9;
      case "HAEC": return 2;
      case "HST": return -10;
      case "HKT": return 8;
      case "HMT": return 5;
      case "HOVST": return 8;
      case "HOVT": return 7;
      case "ICT": return 7;
      case "IDLW": return -12;
      case "IDT": return 3;
      case "IOT": return 3;
      case "IRDT": return 4.5;
      case "IRKT": return 8;
      case "IRST": return 3.5;
      case "IST": return 5.5;
      case "JST": return 9;
      case "KALT": return 2;
      case "KGT": return 6;
      case "KOST": return 11;
      case "KRAT": return 7;
      case "KST": return 9;
      case "LHST": return 10.5; // Lord Howe Standard Time
      case "LINT": return 14;
      case "MAGT": return 12;
      case "MART": return -9.5;
      case "MAWT": return 5;
      case "MDT": return -6;
      case "MEET": return 1;
      case "MEST": return 2;
      case "MHT": return 12;
      case "MIST": return 11;
      case "MIT": return -9.5;
      case "MMT": return 6.5;
      case "MSK": return 3;
      case "MST": return -7;  // Mountain Standard Time (North America)
      case "MUT": return 4;
      case "MVT": return 5;
      case "MYT": return 8;
      case "NCT": return 11;
      case "NDT": return -2.5;
      case "NFT": return 11;
      case "NPT": return 5.75;
      case "NST": return -3.5;
      case "NT": return -3.5;
      case "NUT": return -11;
      case "NZDY": return 13;
      case "NZST": return 12;
      case "OMST": return 6;
      case "ORAT": return 5;
      case "PDT": return -7;
      case "PET": return -5;
      case "PETT": return 12;
      case "PGT": return 10;
      case "PHOT": return 13;
      case "PHT": return 8;
      case "PKT": return 5;
      case "PMDT": return -2;
      case "PMST": return -3;
      case "PONT": return 11;
      case "PST": return -8;  //Pacific Standard Time (North America)
      case "PYST": return -3;
      case "PYT": return -4;
      case "RET": return 4;
      case "ROTT": return -3;
      case "SAKT": return 11;
      case "SAMT": return 4;
      case "SAST": return 2;
      case "SBT": return 11;
      case "SCT": return 4;
      case "SDT": return -10;
      case "SGT": return 8;
      case "SLST": return 5.5;
      case "SRET": return 11;
      case "SST": return 8;  //Singapore Standard Time
      case "SYOT": return 3;
      case "TAHT": return -10;
      case "THA": return 7;
      case "TFT": return 5;
      case "TJT": return 5;
      case "TKT": return 13;
      case "TLT": return 9;
      case "TMT": return 5;
      case "TRT": return 3;
      case "TOT": return 13;
      case "TVT": return 12;
      case "ULAST": return 9;
      case "ULAT": return 8;
      case "UTC": return 0;
      case "UYST": return -2;
      case "UYT": return -3;
      case "UZT": return 5;
      case "VET": return -4;
      case "VLAT": return 10;
      case "VOLT": return 4;
      case "VOST": return 6;
      case "VUT": return 11;
      case "WAKT": return 12;
      case "WAST": return 1;
      case "WAT": return 2;
      case "WEST": return 1;
      case "WET": return 0;
      case "WST": return 8;
      case "YAKT": return 9;
      case "YEKT": return 5;
    }
    return 0;
  } ,
```

The function that calculates the new time from the offset, also using [Date.prototype.getTimezoneOffset()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTimezoneOffset) in order to first calculate UTC time. I need to debug this in detail though.


```
  // Set Time - add Timezone offset
  let nativeUtcOffset = tm.getTimezoneOffset(),
      forceTimeZone = SmartTemplate4.whatIsTimezone; // UTC offset for current time,  in minutes
  if (forceTimeZone)  {
    let forceHours = util.getTimezoneOffset(forceTimeZone);
    timezone = nativeUtcOffset*(-1) + forceHours*60; // offset in minutes!
    util.logDebug("Adding timezone offsets:\n" +
      "UTC Offset: " + nativeUtcOffset/(-60) + 
      "\nForced Timezone[" + forceTimeZone + "]: " + forceHours);
  }
  tm.setTime(time / 1000 + (timezone) * 60 * 1000);

```

**Comment by RealRaven2000** [comment]
Here is another preview version - I change the logic for forcing time zones - effectively it is now subtracting the offset as it should (I tested with GMT and BST and got expected results). I also added the new timezone names CEDT and MESZ.

[smartTemplate-fx-2.2pre17.zip](https://github.com/RealRaven2000/SmartTemplate4/files/3488902/smartTemplate-fx-2.2pre17.zip)

**Comment by RealRaven2000** [comment]
Resolved with Release 2.2 - just submitted to ATN

**Comment by ckeydel** [comment]
Can I still add here or should I open a new issue?

I am at German time (MEST = UTC+0200). The email in TB shows 10:21 AM. In the email header, it is "08:21:05 +0000". In ST4 I have configured X:=timezone(UTC)%...%H%:%M% %date_tz%. When I hit reply I am getting:
12:21 +0200

With timezone(CET) (+0100) it outputs
11:21 +0200

With timezone(MEST) (+0200, my time zone) it outputs
10:21 +0200

With timezone(IDT) (+0300) it outputs
09:21 +0200
In error log to the last one it says "Adding timezone offsets: UTC Offset: 2 hour Forced Timezone Offset: -3 hours Total offset -3600000 ms will be added to time".

The time zone offset is moving in the wrong direction now. Also, the local UTC offset seems to be (still?) ADDED to local time, instead of being SUBSTRACTED to get UTC time. So, your current formula seems to be

<local time> + <local UTC offset> - <forced timezone offset>

when it should be

<local time> - <local UTC offset> + <forced timezone offset>

Sorry I was traveling and couldn't test this earlier.

**Comment by RealRaven2000** [comment]
You can add here - note that SmartTemplate 2.2 was approved on ATN this morning, so back to the drawing board.

Here is a new version, I tested it with setting the timezone to EDT (NewYork time)

tested in Write template:

```
%X:=timezone(EDT)%
<p>datelocal: %datelocal%  EDT timezone</p>
<p>Garrafrauns, %A%, %dateformat(e/n/Y, H:M)%   dateformat EDT timezone</p>
```

tested in reply template:

```
%language(en)%
%X:=timezone(EDT)%
%X:=sent% %A%, %d%.%m%.%Y% %H%:%M% %date_tz%  EDT
```

[smartTemplate-fx-2.2.1pre6.zip](https://github.com/RealRaven2000/SmartTemplate4/files/3492727/smartTemplate-fx-2.2.1pre6.zip)

**Comment by ckeydel** [comment]
Using timezone(UTC) (\<foced timezone offset\>==0) now gives the correct result here:
12.08.2019 08:21

Using MEST (+0200), time is correct, too:
12.08.2019 10:21

Using PST (-0800):
12.08.2019 00:21

These times above are correct according to the hardcoded time zone offsets, but PHOT (-1300) gives:
12.08.2019 21:21

Two problems:
1) The time should be 19:21 (UTC-13 hours). Instead it seems to be \<local time\>-13 hours.
2) The date should be -1 day, so 11.08.2019

The "forward" direction, tested with NZDY (+1300) time zone, seems to work fine, including incrementing the date to the next day.

**Comment by RealRaven2000** [comment]
> 
> 
> Using timezone(UTC) (<foced timezone offset>==0) now gives the correct result here:
> 12.08.2019 08:21
> 
> Using MEST (+0200), time is correct, too:
> 12.08.2019 10:21
> 
> Using PST (-0800):
> 12.08.2019 00:21
> 
> These times above are correct according to the hardcoded time zone offsets, but PHOT (-1300) gives:
> 12.08.2019 21:21
> 
> Two problems:
> 
>     1. The time should be 19:21 (UTC-13 hours). Instead it seems to be <local time>-13 hours.
> 
>     2. The date should be -1 day, so 11.08.2019
> 
> 
> The "forward" direction, tested with NZDY (+1300) time zone, seems to work fine, including incrementing the date to the next day.

Which function did you use? %dateformat( )% or one of the composite ones %a% %d% etc? dateformat should work correctly as it adds the offset to a Date( ) variable which should decrement the day accordingly.

**Comment by ckeydel** [comment]
I had used the composite ones, %H%:%M% etc.

Using both
```
%X:=sent%%X:=timezone(PHOT)%<b>Sent:</b> %A%, %d%.%m%.%Y% %H%:%M%
%X:=sent%%X:=timezone(PHOT)%<b>Sent:</b> %dateformat(A, e/n/Y H:M)%
```
I am getting, even stranger:

```
Sent: Monday, 12.08.2019 21:21
Sent: (Tuesday, 13/8/2019 05:52)
```

**Comment by RealRaven2000** [comment]
Should PHOT set the time forward, since it is a positive number?  CST = -6 (negative) PHOT=13 (positive) the offsets were according to https://en.wikipedia.org/wiki/List_of_time_zone_abbreviations

I think dateformat is possibly not working with X:=sent, I made it to put the current time into a newly written email (no sent header available)

**Comment by ckeydel** [comment]
Sorry, I mixed up the time zones. Trying IDLW (-10 hours), I get (from the same email sent 08:21 UTC resp. 10:21 MEST):
`Sent: Sunday, 11.08.2019 20:21`
So the date is correct, but the time should be 22:21.

**Comment by RealRaven2000** [comment]
Btw you don't need to repeat the timezone zwitch every line, put it at the top. I will test with that now. Maybe I can add "sent" support to dateformat to at least have that in sync; the functions are in different places of the program

**Comment by RealRaven2000** [comment]
OK, I haven't addressed the actual calculation yet, but here is a version that adds X:=Sent support to %dateformat( )% - since my users are going to ask for this anyway. It also doesn't "wrap" the variable into a "clickable" field in that case, but replaces the date right away, just like the other date variables.
 
[smartTemplate-fx-2.2.1pre10.zip](https://github.com/RealRaven2000/SmartTemplate4/files/3493917/smartTemplate-fx-2.2.1pre10.zip)

```
%X:=timezone(IDLW)%
%X:=sent% %A%, %d%/%m%/%Y% %H%:%M%    IDLW <br>
%X:=sent% %dateformat(A, e/n/Y H:M)%  IDLW <br>
```
With this an email sent on 10/08/2019 at 16:00 BST generates:
Saturday, 10/08/2019 03:00 IDLW
Saturday, 10/8/2019 03:00 IDLW

**Comment by RealRaven2000** [comment]
Tried with the same version and email, this time DDUT (10 hours forward)

With this an email sent on 10/08/2019 at 16:00 BST generates:
Sunday, 11/08/2019 01:00 DDUT
Sunday, 11/8/2019 01:00 DDUT

**Comment by ckeydel** [comment]
Yes, you're right - IDWL is -12 hours, not -10, so the time that I thought was wrong was correct in my last post. Date crossing seems to work, too, and so does %dateformat()%. Looking good!

**Comment by RealRaven2000** [comment]
Just released v 2.2.1 on atn. We also allow using %X:=sent% with dateformat( ) now for consistency and ease.

Example:

```
%timezone(GMT)%
%X:=sent%
<p><b>Re: %subject%</b> from %dateformat(A, d/m/Y H:M)%</p>
%X:=today%
<p>Munich, %dateformat(A, d/m/Y H:M)%</p>

```
